Sources/StreamChatUI/Utils/ViewPaginationHandling/StatefulScrollViewPaginationHandler.swift

```diff
@@ -0,0 +1,75 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import UIKit
+
+/// A component responsible to handle when to load new pages in a scrollView holding state associated to the content view
+final class StatefulScrollViewPaginationHandler: StatefulViewPaginationHandling {
+    private var bottomPageRequestItemCount: Int?
+    private var topPageRequestItemCount: Int?
+
+    var canRequestNewTopPage: Bool {
+        topPageRequestItemCount == nil
+    }
+
+    var canRequestNewBottomPage: Bool {
+        bottomPageRequestItemCount == nil
+    }
+
+    private weak var scrollView: UIScrollView?
+    private var observation: NSKeyValueObservation?
+    private var previousPosition: CGFloat = 0.0
+
+    var topThreshold: Int = 100
+    var bottomThreshold: Int = 400
+
+    var onNewTopPage: StatefulViewPaginationHandlingBlock?
+    var onNewBottomPage: StatefulViewPaginationHandlingBlock?
+
+    init(scrollView: UIScrollView) {
+        self.scrollView = scrollView
+
+        observation = self.scrollView?.observe(\.contentOffset, changeHandler: { [weak self] scrollView, _ in
+            self?.onChanged(scrollView)
```

Comment:
this will be called a lot. Do you think it might have impact on performance?

Reply:
It's a shame we can't use `scrollView.publisher(for: \.contentOffset)` and debounce it :/

Reply:
why not use it though? We could put iOS 13 check, and use either the observation or the combine subscription. We anyway might get rid of iOS 11 (we're waiting on the usage data from product).

Reply:
This is the same code we have in the current implementation of ScrollViewPaginationHandler 😅 . Did not want to create yet another intermediate class given it is not much code that can be shared.

Reply:
> why not use it though? We could put iOS 13 check, and use either the observation or the combine subscription. We anyway might get rid of iOS 11 (we're waiting on the usage data from product).

It is not only about adding an availability check for iOS 13, you need conditional compiler annotations to allow the usage of AnyCancellable. It can be done, but 🤷

Reply:
The cancellable can be declared as `Any`, but it all depends if there's possible performance impact here. If there isn't, then it's not worth it imo.

Reply:
If we are sure that Combine was released initially with all those APIs, we could. I am just not entirely sure about it

Reply:
The documentation states iOS 13.0+ for both, so it should be the case 🤔 But `import Combine` does not raise any warnings, while it does when using its APIs

Reply:
I think the main question here is: do you think there's a performance impact? Maybe just a quick instruments run with / without this call can verify this. 
If there isn't, then I don't see value in debouncing and changing the implementation.

Reply:
No big performance hit, but debouncer would be preferrable